Post: small post about "latest core requirements" (#226)
readthedocs/blog: python-core-requirements-changed.rst

> .. post:: July 7, 2023

Author: ```suggestion
.. post:: July 10, 2023
```

> .. post:: July 7, 2023
>    :tags: builders
>    :author: Manuel
>    :location: BCN
>    :category: Changelog
> 
> Python "core requirements" installed by default have changed
> ============================================================
> 
> Starting on **August 7, 2023** all new projects imported on Read the Docs
> will install only ``sphinx``, ``mkdocs`` and ``readthedocs-sphinx-ext`` as "core requirements".
> Besides, the new behavior will install the latest version of these requirements.
> 
> Note that previously Read the Docs was installing also
> ``jinja``, ``sphinx-rtd-theme``, ``pillow``, ``mock``, ``alabaster``, ``commonmark`` and ``recommonmark``
> specifying particular versions depending on different factors that were confusing for users and hard to debug.
> 
> We are moving away from executing commands and installing dependencies on behalf of the users
> and trying to make the build experience on Read the Docs the same as building locally,
> without surprises.
> 
> Projects imported previous to August 7, 2023 will keep the actual behavior and won't be affected by this change.
> If you have a project you want to install the latest version of these requirements,
> you can use a ``requirements.txt`` file to specify their versions.
> Read more about :doc:`how to use a requirements.txt file in our documentation <readthedocs:config-file/v2>`.

Reviewer: ```suggestion
.. post:: July 7, 2023
   :tags: builders
   :author: Manuel
   :location: BCN
   :category: Changelog

Python "core requirements" for new projects will install latest version
=======================================================================

Starting on **August 7, 2023** all new projects imported on Read the Docs
will install only ``sphinx``, ``mkdocs`` and ``readthedocs-sphinx-ext`` as "core requirements".
The default behavior will be to install the latest version of these requirements.

Note that previously Read the Docs was installing also
``jinja``, ``sphinx-rtd-theme``, ``pillow``, ``mock``, ``alabaster``, ``commonmark`` and ``recommonmark``
specifying particular versions depending on different factors that were confusing for users and hard to debug.

We are moving away from executing commands and installing dependencies on behalf of the users
and trying to make the build experience on Read the Docs the same as building locally,
without surprises.

Projects imported before August 7, 2023 will keep the existing behavior and won't be affected by this change.
If you have a project you want to install the latest version of these requirements,
you can use a ``requirements.txt`` file to specify their versions, or no version to automatically install the latest version.
Read more about :doc:`how to use a requirements.txt file in our documentation <readthedocs:config-file/v2>`.
```

Author: > If you have a project you want to install the latest version of these requirements,
you can use a ``requirements.txt`` file to specify their versions, or no version to automatically install the latest version.

I thought about this as well, but this is not true:

```
$ pip install "sphinx<2.2"
...
$ echo "sphinx" > requirements.txt
$ pip install -r requirements.txt
```

Nothing was upgraded at this point. However, projects with `PIP_ALWAYS_UPGRADE` will add `--upgrade` to that command:

```
$ pip install --upgrade -r requirements.txt
```

and the latest Sphinx will be installed.

Author: This code lives at https://github.com/readthedocs/readthedocs.org/blob/78751790cdecd22c635611b51b35779022a0ccd4/readthedocs/doc_builder/python_environments.py#L296-L297

We only have 5 projects on .org with this feature flag enabled (https://admin.ops.readthedocs.org:10000/admin/projects/feature/5/change/). Does it make sense to apply it for all the projects? Man, we have a lot of feature flags that make our workflow too complex 😞 

I think it makes sense, because it's aligned with what the user wants, right? If they specify `sphinx` without versions it's because they want the latest one.

Author: On .com we don't even have that feature flag created, so, 0 projects are using it.

Author: Pip documentation about `--upgrade`: https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-U

Reviewer: It sounds like users can set the ENV var to have pip always upgrade? If that works, I'm fine with that until users can call pip with whatever flags they want in `build.jobs.install`.

Author: That's a pretty good idea actually. I wasn't aware this can be done with environment variables. I think it would be really good to document/suggest this for projects with this requirement 👍
